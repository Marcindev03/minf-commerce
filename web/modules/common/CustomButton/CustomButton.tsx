import classNames from "classnames";
import Link from "next/link";
import { FC, ReactNode } from "react";
import { HiArrowPath } from "react-icons/hi2";

type CustomButtonProps = {
  href?: string;
  className?: string;
  variant?: "primary" | "outline";
  disabled?: boolean;
  isLoading?: boolean;
  loadingText?: string;
  submitButton?: boolean;
  onClick?: () => void;
  children: ReactNode;
};

export const CustomButton: FC<CustomButtonProps> = ({
  href,
  className,
  variant = "primary",
  disabled,
  isLoading = false,
  loadingText,
  submitButton = false,
  onClick,
  children,
}) => {
  const Button = () => (
    <button
      className={classNames(" py-2 px-5 border rounded", className, {
        "text-white border-blue-500 bg-blue-500 hover:bg-blue-800 hover:border-blue-800":
          variant === "primary",
        "text-black bg-white border-blue-400 hover:bg-blue-400 hover:text-white":
          variant === "outline",
        "cursor-not-allowed opacity-50": disabled,
      })}
      onClick={onClick}
      disabled={disabled || isLoading}
      type={submitButton ? "submit" : "button"}
    >
      <>
        {isLoading ? (
          <div className="w-full flex justify-center">
            {loadingText}
            <HiArrowPath
              color="white"
              size="1.5rem"
              className="animate-spin ml-4"
              data-testid="custom_button_loading_icon"
            />
          </div>
        ) : (
          children
        )}
      </>
    </button>
  );

  if (href) {
    return (
      <Link href={href}>
        <Button />
      </Link>
    );
  }

  return <Button />;
};
