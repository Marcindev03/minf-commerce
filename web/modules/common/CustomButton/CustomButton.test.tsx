import { fireEvent, render, screen } from "@testing-library/react";
import renderer from "react-test-renderer";
import { CustomButton } from "./CustomButton";

const buttonText = "Click me!";
const buttonLoadingText = "loading...";
const href = "/products";
const className = "py-10";
const buttonPrimaryStyles =
  "text-white border-blue-500 bg-blue-500 hover:bg-blue-800 hover:border-blue-800";
const buttonOutlineStyles =
  "text-black bg-white border-blue-400 hover:bg-blue-400 hover:text-white";
const buttonDisabledStyles = "cursor-not-allowed opacity-50";

describe("Custom button", () => {
  test("should render children", () => {
    render(<CustomButton>{buttonText}</CustomButton>);

    const button = screen.getByRole("button");
    expect(button).toHaveTextContent(buttonText);
  });

  test("should call onClick prop when called", () => {
    const handleClick = jest.fn();
    render(<CustomButton onClick={handleClick}>{buttonText}</CustomButton>);

    const button = screen.getByRole("button");
    fireEvent.click(button);
    expect(handleClick).toBeCalledTimes(1);
  });

  test("should be disabled when disabled props is provided", () => {
    render(<CustomButton disabled>{buttonText}</CustomButton>);

    const button = screen.getByRole("button");
    expect(button).toBeDisabled();
  });

  test("should not fire onClick when disabled", () => {
    const handleClick = jest.fn();
    render(
      <CustomButton disabled onClick={handleClick}>
        {buttonText}
      </CustomButton>
    );

    const button = screen.getByRole("button");

    fireEvent.click(button);

    expect(button).toBeDisabled();
    expect(handleClick).not.toHaveBeenCalled();
  });

  test("should apply disabled styles when disabled prop is provided", () => {
    render(<CustomButton disabled>{buttonText}</CustomButton>);

    const button = screen.getByRole("button");
    expect(button.className).toContain(buttonDisabledStyles);
  });

  test("should render a tag when href is provided", () => {
    render(<CustomButton href={href}>{buttonText}</CustomButton>);

    const link = screen.getByRole("link");
    expect(link).toBeInTheDocument();
    expect(link).toHaveAttribute("href", href);
  });

  test("should not render a tag whe href is not provided", () => {
    render(<CustomButton>{buttonText}</CustomButton>);

    const link = screen.queryByRole("link");
    expect(link).not.toBeInTheDocument();
  });

  test("should should add provided className to existing ones", () => {
    render(<CustomButton className={className}>{buttonText}</CustomButton>);

    const button = screen.getByRole("button");
    expect(button.className.split(" ")).toContain(className);
  });

  test("should apply primary styles defaulty", () => {
    render(<CustomButton>{buttonText}</CustomButton>);

    const button = screen.getByRole("button");
    expect(button.className).toContain(buttonPrimaryStyles);
  });

  test("should apply outline styles when variant is outline", () => {
    render(<CustomButton variant="outline">{buttonText}</CustomButton>);

    const button = screen.getByRole("button");
    expect(button.className).toContain(buttonOutlineStyles);
  });

  test("should have button type defaulty ", () => {
    render(<CustomButton>{buttonText}</CustomButton>);

    const button = screen.getByRole("button");
    expect(button).toHaveAttribute("type", "button");
  });

  test("should have submit type when submitButton prop is provided", () => {
    render(<CustomButton submitButton>{buttonText}</CustomButton>);

    const button = screen.getByRole("button");
    expect(button).toHaveAttribute("type", "submit");
  });

  test("should render loading when isLoading prop is true", () => {
    render(<CustomButton isLoading>{buttonText}</CustomButton>);

    const loadingIcon = screen.getByTestId("custom_button_loading_icon");
    expect(loadingIcon).toBeInTheDocument();
  });

  test("should render loadingText when loadingText prop is true", () => {
    render(
      <CustomButton isLoading loadingText={buttonLoadingText}>
        {buttonText}
      </CustomButton>
    );

    const loadingElement = screen.getByText(buttonLoadingText);
    expect(loadingElement).toBeInTheDocument();
  });

  test("should not render loadingText when isLoading prop is not passed", () => {
    render(
      <CustomButton loadingText={buttonLoadingText}>{buttonText}</CustomButton>
    );

    const loadingTextElement = screen.queryByText(buttonLoadingText);
    expect(loadingTextElement).toBeNull();
  });

  test("should match snapshot", () => {
    const handleClick = jest.fn();
    const tree = renderer
      .create(
        <CustomButton
          disabled
          onClick={handleClick}
          href={href}
          className={className}
        >
          {buttonText}
        </CustomButton>
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
