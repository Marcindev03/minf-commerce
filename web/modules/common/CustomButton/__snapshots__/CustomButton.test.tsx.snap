// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Custom button should match snapshot 1`] = `
<a
  href="/products"
  onClick={[Function]}
  onMouseEnter={[Function]}
  onTouchStart={[Function]}
>
  <button
    className="bg-blue-500 py-3 px-5 text-white rounded hover:bg-blue-800 py-10"
    disabled={true}
    onClick={[MockFunction]}
  >
    Click me!
  </button>
</a>
`;
